feat: receive and handle nr+ data from nRF Cloud message routing service

Lenakh97 opened the pull request.

Lenakh97 committed:
wip

Lenakh97 committed:
wip

coderbyheart commented on cdk/stacks/NRPlusDemoStack.ts:
```diff
@@ -0,0 +1,61 @@
+			lambdaSources.webhookHandler,
+			{
+				runtime: Lambda.Runtime.NODEJS_22_X,
+				description: '',
```
Please give each lambda a helpful descriptions, this is useful when locating them in the AWS console.

coderbyheart commented on cdk/stacks/NRPlusDemoStack.ts:
```diff
@@ -0,0 +1,61 @@
+				initialPolicy: [
+					new IAM.PolicyStatement({
+						actions: ['iot:UpdateThingShadow'],
+						resources: ['*'],
```
Should be limited to shadow with name lwm2m only.

Lenakh97 replied:
Does it look good now? Or do we want more '*' in the name?

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,175 @@
+	)
+	.handler(async (event, context): Promise<APIGatewayProxyResultV2> => {
+		console.log('context:', context.validInput)
+		if (event.requestContext.http.method === 'POST') {
```
**style:** keep code left, return early.

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,175 @@
+			}
+			return {
+				statusCode: 200,
+				body: JSON.stringify({ message: 'Received' }),
```
Is the body response required? If not, remove.

coderbyheart replied:
... if yes, send content-type header.

coderbyheart commented on package.json:
```diff
@@ -26,36 +26,36 @@
+    "@bifravst/aws-cdk-ecr-helpers": "1.3.100",
+    "@bifravst/aws-cdk-lambda-helpers": "3.5.1",
+    "@bifravst/eslint-config-typescript": "6.4.3",
+    "@bifravst/prettier-config": "1.1.12",
```
**nitpick:** unrelated change

Lenakh97 committed:
wip

Lenakh97 committed:
wip

Lenakh97 committed:
wip

Lenakh97 committed:
wip

Lenakh97 committed:
wip

Lenakh97 commented:
Now added logic for updating thing if it's not existing, og some tests around it. Also added a test for input validation with real data. @coderbyheart 

Some questions/concerns:
- thingExists.ts -> Not sure what the best way to do this. I believe either change the function to make sure we always return false if there is an error or handle it differently if there is an error. The idea was that the handler doesn't need to know about the error - only if it exists or not. And that's why I tried handling it inside the function and not in the handler. Any input?
- I'm using 'continue' in the if statements, because as far as I know there is never a shadow update + location (CoAP) update in one message. Not sure if this is ideal, in case I'm wrong and/or this changes? 
- The conditions in the if sentences: I feel like there should be an easier way to check this, but I've gotten som other problems when doing this/ts was not happy.

Lenakh97 committed:
wip

coderbyheart commented on cdk/resources/WebsocketAPI.ts:
```diff
@@ -340,6 +340,32 @@ export class WebsocketAPI extends Construct {
+					actions: [
+						{
+							lambda: {
+								functionArn: publishToWebsocketClients.functionArn,
```
Please make a copy of the lambda so the shadow update is not sent with the `@context` `https://thingy.rocks/device-shadow` but `https://thingy.rocks/lwm2m-shadow` so it aligns with the onMessage lanbda: https://github.com/NordicPlayground/thingy-rocks-cloud-aws-js/blob/05f63623d396d246f7f60ab8fa9b5c9bd29da13c/lambda/onMessage.ts#L131

Lenakh97 replied:
I'm not a 100% sure what to do to send it with the correct context? I can't find exactly where this happens and how to choose the correct context.

coderbyheart replied:
This is set here: https://github.com/NordicPlayground/thingy-rocks-cloud-aws-js/blob/50eab172778e1a7c382769415b5b6dfc506ab902/lambda/notifyClients.ts#L100

You need a new lambda that correctly sets the `@context` in the call to `notifyClients()`

Lenakh97 replied:
**Question**: I am getting both 'device-shadow' & 'lwm2m-shadow' in the frontend from the nr+ devices. It looks like the onMessage function catches the updates from the devices I have added. Do I need to add a new lambda to update it via the 'publishToWebsocketClients' lambda - or can I delete what I added in 'websocketAPI.ts' since it's handled in the 'onMessage' function? As far as I can see the onMessage handler gets all the deltas from the 'lwm2m' named shadows?

coderbyheart replied:
The `lwm2m-shadow` is automatically sent because of [this rule](https://github.com/NordicPlayground/thingy-rocks-cloud-aws-js/blob/441b0b679229d7b8c98b0e116dac74b06ac5a7f6/cdk/resources/LwM2MObjectsHistory.ts#L77). So, you are correct, you do not need to send the updates you write, it is picked up automatically.

coderbyheart commented on cdk/stacks/NRPlusDemoStack.ts:
```diff
@@ -0,0 +1,68 @@
+							'iot:CreateThing',
+						],
+						resources: [
+							'arn:aws:iot:eu-central-1:374216331074:thing/*/shadow/lwm2m',
```
```suggestion
							`arn:aws:iot:${Stack.of(this).region}:${Stack.of(this).account}:thing/*/shadow/lwm2m`,
```

Do not hardcode account and region in CDK.

coderbyheart commented on cdk/stacks/NRPlusDemoStack.ts:
```diff
@@ -0,0 +1,68 @@
+		new CfnOutput(this, 'WebhookFunctionURL', {
+			exportName: 'WebhookFunctionURL',
+			description: 'The URL for the webhook function',
+			value: `${functionUrl.url}`,
```
```suggestion
			value: functionUrl.url,
```

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,165 @@
+				body: 'Method Not Allowed',
+			}
+		}
+		const validatedInput = context.decodedInput
```
Extract this logic in a separate function and add a test.

Lenakh97 replied:
What logic? Statuscode 405 handling? or line 118?

coderbyheart replied:
The entire logic from line 118 to 161

coderbyheart commented on nordicNRPlus/Thingexists.spec.ts:
```diff
@@ -0,0 +1,63 @@
+				thingArn: '',
+				thingId: '',
+				thingName: '',
+				thingTypeName: '',
+				version: 3,
+			}),
```
Simplify this to only provide the necessary data:
```suggestion
			Promise.resolve({
				attributes: { name: 'myDevice123' },
			}),
```

coderbyheart commented on nordicNRPlus/locationDataFromCOAPToLwm2m.spec.ts:
This is tested in https://github.com/NordicPlayground/thingy-rocks-cloud-aws-js/blob/2e39a80e978479fa65a46c88adab0dddabfb9bbb/proto-asset_tracker_v2+AWS/examples/examples.spec.ts#L39-L74

coderbyheart replied:
so you could re-use the `transformShadowUpdateToLwM2M()` method.

Lenakh97 replied:
I don't have a shadow with the location data so I'm not sure I can actually use this function?

coderbyheart replied:
Ah, right, this is the the legacy nRF Cloud message, not the shadow.

coderbyheart commented on nordicNRPlus/parseCbor.ts:
```diff
@@ -0,0 +1,27 @@
+	uncertainty: number
+	src: string
+}
+export const parseCbor = (cborData: string): LocationData => {
```
```suggestion
export const parseCBORLocationData = (cborData: string): LocationData => {
```

Specific implementation for just one object type.

coderbyheart commented on nordicNRPlus/thingExists.ts:
```diff
@@ -0,0 +1,17 @@
+		await iotClient.send(new DescribeThingCommand({ thingName }))
+		return true
+	} catch (err: any) {
+		if (err.name === 'ResourceNotFoundException') {
+			return false
+		} else {
+			return false
+		}
```
```suggestion
		return false
```

Both paths return the same.

coderbyheart commented on nordicNRPlus/validateInput.spec.ts:
```diff
@@ -0,0 +1,25 @@
+import { describe, it } from 'node:test'
+import { inputSchemaLwm2mMessage } from '../lambda/webhookHandler.ts'
+
+void describe('validateExampleInput', () => {
```
```suggestion
void describe('inputSchemaLwm2mMessage', () => {
```

coderbyheart commented:
thingExists.ts: yeah that's OK if its another error there is nothing you can do anyway. But I would throw an unexpected error so it shows up in the logs.

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,165 @@
+			const { teamId, deviceId } = message
+			const thingName = `${teamId}-${deviceId}`
+			//check if thing exists, if not create it
+			if ((await thingExists(iotClient, thingName)) === false) {
```
```suggestion
			if ((await thingExists(thingName)) === false) {
```

Do not pass dependencies on every call.

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,165 @@
+			}
+		}
+		const validatedInput = context.decodedInput
+		for (const message of validatedInput.messages) {
```
Change it to "wishful thinking":

```typescript
await ensureThing(thingName)
const maybeProcessedMessage = processMessage(message, thingName)
if (!maybeProcessedMessage) console.log('Unhandled message shape:', message)
```

Lenakh97 replied:
Is it better now? Not as clean as your suggestion, since I have a try/catch. Also updates the shadow in the handler - should that be in the 'processMessage' function?

coderbyheart replied:
Yes, so it can be easier tested

coderbyheart commented:
Once you have extracted the method from the lambda we can look into how to refactor it and make it more modular.

The core idea is to use a list of functions and a higher-order function (like reduce or fold) to process the data through them sequentially.

---

### **The Functional Approach: A Pipeline with Option/Maybe**

Instead of objects passing a request to the next object in a chain, you create a list of functions. You then iterate through the list, applying each function to the data until one of them successfully handles it.

The key is how a function signals "I can't handle this." This is a perfect use case for the **Option** type (also known as Maybe in languages like Haskell).

* If a function can handle the data, it returns Some(result).  
* If it cannot, it returns None (or Nothing).

The pipeline then stops as soon as it receives the first Some value.

---

### **A Simple Example**

Let's imagine you have data and you want to process it. You have three functions: one that handles integers, one that handles short strings, and one for long strings.

#### **1. Define the Handler Functions**

Each function takes the data and returns an Option of the result.

```
// Handles only integers  
function handleIntegers(data):  
  if is_integer(data):  
    return Some("Processed an integer: " + data)  
  else:  
    return None

// Handles strings with less than 5 characters  
function handleShortStrings(data):  
  if is_string(data) and length(data) < 5:  
    return Some("Processed a short string: " + data)  
  else:  
    return None

// A fallback for other strings  
function handleAnyString(data):  
  if is_string(data):  
    return Some("Processed some other string: " + data)  
  else:  
    return None
```

#### **2. Create the Chain (A List of Functions)**

This is your "chain of responsibility."

```
const handlers = [  
  handleIntegers,  
  handleShortStrings,  
  handleAnyString  
]
```

#### **3. The Executor**

You can use a reduce or a simple loop to execute the pipeline. The reduce approach is often considered more idiomatic in functional programming.

```
function process(data, chain):  
  // Start with a "not handled" state (None)  
  // The 'reduce' function will try each handler until one returns 'Some'  
  return chain.reduce((result, handler) => {  
    // If we already have a result, just pass it along  
    if result is Some:  
      return result  
    // Otherwise, try the next handler in the chain  
    else:  
      return handler(data)  
  }, None) // Initial value is None
```

#### **4. Running the Code**

```
process(123, handlers)       // Returns: Some("Processed an integer: 123")  
process("Hi", handlers)        // Returns: Some("Processed a short string: Hi")  
process("Hello World", handlers) // Returns: Some("Processed some other string: Hello World")  
process(true, handlers)      // Returns: None (no handler could process a boolean)
```

Lenakh97 committed:
Update cdk/stacks/NRPlusDemoStack.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/Thingexists.spec.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update cdk/stacks/NRPlusDemoStack.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/thingExists.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/parseCbor.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/validateInput.spec.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
wip

Lenakh97 commented:
> Once you have extracted the method from the lambda we can look into how to refactor it and make it more modular.
> 
> The core idea is to use a list of functions and a higher-order function (like reduce or fold) to process the data through them sequentially.
> 
> ### **The Functional Approach: A Pipeline with Option/Maybe**
> Instead of objects passing a request to the next object in a chain, you create a list of functions. You then iterate through the list, applying each function to the data until one of them successfully handles it.
> 
> The key is how a function signals "I can't handle this." This is a perfect use case for the **Option** type (also known as Maybe in languages like Haskell).
> 
> * If a function can handle the data, it returns Some(result).
> * If it cannot, it returns None (or Nothing).
> 
> The pipeline then stops as soon as it receives the first Some value.
> 
> ### **A Simple Example**
> Let's imagine you have data and you want to process it. You have three functions: one that handles integers, one that handles short strings, and one for long strings.
> 
> #### **1. Define the Handler Functions**
> Each function takes the data and returns an Option of the result.
> 
> ```
> // Handles only integers  
> function handleIntegers(data):  
>   if is_integer(data):  
>     return Some("Processed an integer: " + data)  
>   else:  
>     return None
> 
> // Handles strings with less than 5 characters  
> function handleShortStrings(data):  
>   if is_string(data) and length(data) < 5:  
>     return Some("Processed a short string: " + data)  
>   else:  
>     return None
> 
> // A fallback for other strings  
> function handleAnyString(data):  
>   if is_string(data):  
>     return Some("Processed some other string: " + data)  
>   else:  
>     return None
> ```
> 
> #### **2. Create the Chain (A List of Functions)**
> This is your "chain of responsibility."
> 
> ```
> const handlers = [  
>   handleIntegers,  
>   handleShortStrings,  
>   handleAnyString  
> ]
> ```
> 
> #### **3. The Executor**
> You can use a reduce or a simple loop to execute the pipeline. The reduce approach is often considered more idiomatic in functional programming.
> 
> ```
> function process(data, chain):  
>   // Start with a "not handled" state (None)  
>   // The 'reduce' function will try each handler until one returns 'Some'  
>   return chain.reduce((result, handler) => {  
>     // If we already have a result, just pass it along  
>     if result is Some:  
>       return result  
>     // Otherwise, try the next handler in the chain  
>     else:  
>       return handler(data)  
>   }, None) // Initial value is None
> ```
> 
> #### **4. Running the Code**
> ```
> process(123, handlers)       // Returns: Some("Processed an integer: 123")  
> process("Hi", handlers)        // Returns: Some("Processed a short string: Hi")  
> process("Hello World", handlers) // Returns: Some("Processed some other string: Hello World")  
> process(true, handlers)      // Returns: None (no handler could process a boolean)
> ```

Tried doing something like this in the 'processMessage' function. Is that how you imagined it? It's all in one file for now, not sure if it should stay like that. Also not completely sure where the handlers should be introduced. Inside the 'processMessage' function? Or do I need another function that combines the handlers and the processing?

Lenakh97 committed:
wip

Lenakh97 committed:
wip

coderbyheart commented on lambda/ensureThingExists.ts:
```diff
@@ -0,0 +1,15 @@
+				}),
+			)
+		}
+	}
```
```suggestion

export const ensureThingExists =
	(iotClient: IoTClient, thingExists: EnsureThingFn) => {
	
	const e = thingExists(iotClient)
	
	return async (thingName: string): Promise<void> => {
	    if (await e(thingName)) return
		await iotClient.send(
			new CreateThingCommand({
				thingName,
				thingTypeName: 'nordic-nrplus',
			}),
		)
	}
	
	}
```

1. Do not inject dependencies on every call
2. Keep code left

coderbyheart commented on nordicNRPlus/processNrplusMessagesAndUpdateThingShadow.ts:
```diff
@@ -0,0 +1,41 @@
+import type { Static } from '@sinclair/typebox'
+import type { inputSchemaLwm2mMessage } from '../lambda/webhookHandler.ts'
```
Move this here, so you do not include the lambda on unit tests, because lambdas access global state and do other stuff (e.g.: 

```typescript
export const iotData = new IoTDataPlaneClient({})
const iotClient = new IoTClient({})
```

Lenakh97 replied:
Do you mean moving all of the schemas to this file? Is it better that the lambda imports the schema from this file than the opposite?

coderbyheart replied:
Yes, the lambda should import them.

coderbyheart commented on lambda/webhookHandler.ts:
```diff
@@ -0,0 +1,125 @@
+export const handler = middy<APIGatewayProxyEventV2, APIGatewayProxyResultV2>()
+	.use(inputOutputLogger())
+	.use(validateInput(inputSchemaLwm2mMessage))
+	.handler(async (event, context): Promise<APIGatewayProxyResultV2> => {
```
```suggestion
	.handler(async (event, context) => {
```

Typed via middy already

coderbyheart commented on nordicNRPlus/locationDataFromCOAPToLwm2m.ts:
```diff
@@ -0,0 +1,22 @@
+			},
+		},
+	]
+}
```
```suggestion
export const locationDataFromCOAPToLwm2m = (
	location: LocationData,
	ts: number,
): LwM2MObjectInstance[] => [
		{
			ObjectID: 14201,
			ObjectVersion: '1.0',
			ObjectInstanceID: location.src === 'WIFI' ? 1 : 0,
			Resources: {
				'0': location.lat,
				'1': location.lon,
				'3': location.uncertainty,
				'6': location.src,
				'99': ts / 1000,
			},
		},
	]
```

coderbyheart commented on nordicNRPlus/processNrplusMessagesAndUpdateThingShadow.spec.ts:
```diff
@@ -0,0 +1,194 @@
+				ObjectVersion: '1.0',
+				Resources: { '0': 696113427, '1': 0, '99': 1760524760 },
+			},
+		]) // updateShadow is called with correct message
```
```suggestion
		], 'updateShadow is not called with correct message')
```

coderbyheart commented on nordicNRPlus/processNrplusMessagesAndUpdateThingShadow.ts:
```diff
@@ -0,0 +1,41 @@
+			const { teamId, deviceId } = message
+			const thingName = `${teamId}-${deviceId}`
+			try {
+				await ensureThing(thingName)
```
Do not call this for every message, remember if the thing exists.

Lenakh97 replied:
https://github.com/NordicPlayground/thingy-rocks-cloud-aws-js/pull/18/commits/ef4d377b58d6a3e548942f11471e70e9bea6a140

coderbyheart commented on nordicNRPlus/thingExists.ts:
```diff
@@ -0,0 +1,13 @@
+	try {
+		await iotClient.send(new DescribeThingCommand({ thingName }))
+		return true
+	} catch (err: any) {
```
```suggestion
	} catch {
```

coderbyheart commented on nordicNRPlus/thingExists.ts:
```diff
@@ -0,0 +1,13 @@
+export const thingExists = async (
+	iotClient: IoTClient,
+	thingName: string,
+): Promise<boolean> => {
```
```suggestion
export const thingExists = (
	iotClient: IoTClient) => async (
	thingName: string,
): Promise<boolean> => {
```

Bind dependencies once.

Lenakh97 committed:
Update nordicNRPlus/thingExists.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/thingExists.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update nordicNRPlus/locationDataFromCOAPToLwm2m.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update lambda/ensureThingExists.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update lambda/webhookHandler.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
wip

Lenakh97 committed:
wip

Lenakh97 committed:
wip

coderbyheart commented on nordicNRPlus/processNrplusMessagesAndUpdateThingShadow.ts:
```diff
@@ -0,0 +1,132 @@
+				})
+				ensuredThings.set(thingName, ensurePromise)
+			}
+			await ensurePromise
```
**suggestion**: refactor this and move the caching logic outside of this function, so it continues to call `ensureThing()`, however you pass in a memoizing function. That allows you to also separate the unit tests.

Lenakh97 committed:
wip
